File: algorithms/java/106_construct_binary_tree_from_postorder_and_inorder_traversal/Construct.java
public class Construct {
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        if (inorder == null || inorder.length == 0 || postorder == null || postorder.length == 0) return null;
        int[] inIndex = new int[]{inorder.length - 1};
        int[] postIndex = new int[]{postorder.length - 1};
        return construct(inorder, postorder, inIndex, postIndex, Integer.MAX_VALUE);    
    }
    // left right root
    private TreeNode construct(int[] inorder, int[] postorder, int[] inIndex, int[] postIndex, int nextInorderValue) {
        if (postIndex[0] < 0 || inorder[inIndex[0]] == nextInorderValue) return null;
        TreeNode root = new TreeNode(postorder[postIndex[0]]);
        postIndex[0]--;
        root.right = construct(inorder, postorder, inIndex, postIndex, root.val);
        inIndex[0]--;
        root.left = construct(inorder, postorder, inIndex, postIndex, nextInorderValue);
        return root;
    }
}

File: algorithms/cpp/106_in_po_construct/Construct.cpp
class Construct {
public:
	TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        int in_index = inorder.size() - 1;
        int po_index = postorder.size() - 1;
        return construct(inorder, in_index, postorder, po_index, INT_MAX);
    }
private:
    TreeNode* construct(vector<int>& inorder, int& in_index, vector<int>& postorder, int& po_index, int next_in_val) {
        if (po_index < 0 || inorder[in_index] == next_in_val)
            return NULL;
        TreeNode* root = new TreeNode(postorder[po_index]);
        --po_index;
        root->right = construct(inorder, in_index, postorder, po_index, root->val);
        --in_index;
        root->left = construct(inorder, in_index, postorder, po_index, next_in_val);
        return root;
    }
};
